SI-4625 Recognize App in script

som-snytt opened the pull request:
Cheap name test: if the script object extends "App",
take it for a main-bearing parent.

Note that if `-Xscript` is not `Main`, the default,
then the source is taken as a snippet and there is
no attempt to locate an existing `main` method.

som-snytt committed:
SI-4625 Recognize App in script

Cheap name test: if the script object extends "App",
take it for a main-bearing parent.

Note that if `-Xscript` is not `Main`, the default,
then the source is taken as a snippet and there is
no attempt to locate an existing `main` method.

retronym commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -380,11 +380,15 @@ self =>
           case _                                        => false
         }
+        def isApp(t: Tree) = t match {
+          case Template(ps, _, _) => ps.exists { case Ident(x) if x.decoded == "App" => true ; case _ => false }
```
`x == tpnme.App` would be more idiomatic. That form avoids needless string creation.

retronym replied:
or even `case Ident(tpnme.App) =>`

som-snytt replied:
Thx, I grepped for it but missed it, trying to beat the evening rush home. Unless you mean add it to names.

retronym replied:
Yep, I mean to add it.

retronym commented:
LGTM, nitpick aside.

retronym commented:
Scope creep request: Would be great if we could detect Appy modules in files with more than one definition.

```
% cat sandbox/test.scala && scala sandbox/test.scala
object Test {  def main(args: Array[String]) { println("!!") }}
!!

% cat sandbox/test.scala && scala sandbox/test.scala
object O
object Test {  def main(args: Array[String]) { println("!!") }}

%
```

som-snytt commented:
I considered a REPLish strategy, put the entire text in a package object.

som-snytt committed:
SI-4625 App is a thing

Scripting knows it by name.

som-snytt committed:
SI-4625 Permit arbitrary top-level in script

In an unwrapped script, where a `main` entry point is discovered
in a top-level object, retain all top-level classes.

Everything winds up in the default package.

som-snytt committed:
SI-4625 Warn when discarding script object

It's pretty confusing when your script object becomes a local
and then nothing happens. Such as when you're writing a test and
it takes forever to figure out what's going on.

som-snytt committed:
SI-4625 Warn on first non-toplevel only

Fixed the warning when main module is accompanied by snippets.

Minor clean-up so even I can follow what is returned.

som-snytt commented:
I thought the scope creep was the guy with binoculars who peers in people's windows.

lrytz commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -407,54 +407,58 @@ self =>
       }
 
-      if (mainModuleName == newTermName(ScriptRunner.defaultScriptMain))
-        searchForMain() foreach { return _ }
```
wow, never seen that pattern before..!

lrytz commented:
LGTM, too

som-snytt commented:
I just expected this to work, but I guess it hasn't been merged forward yet...
